### PR TITLE
Execute complex request hierarchies

### PR DESCRIPTION
**Pull request by danielmitterdorfer**
With this commit we add the ability to Rally to execute arbitrarily complex
request hierarchies in benchmarks. This is implemented with a new `composite`
operation type. It allows to nest multiple so-called streams which are executed
concurrently. At the moment, the request hierarchy needs to be specified inline
in the track. It is recommended to store more complex request hierarchies in a
separate file and include it with Jinja's `include` directive.

Apart from this change, we also align the internal mapping of operation-type
names to `OperationType` enumeration constants. This is mainly done to simplify
usage and implementation of the `Composite` runner as it can directly use the
operation-types that are specified in the track rather than converting them
internally.

### CODE REVIEW

**Comment by danielmitterdorfer** [comment]
Note to reviewers: I have added the following snippet to test a more complex operation in the geonames track:

```json
{
          "name": "load-geonames-search-page",
          "warmup-iterations": 200,
          "iterations": 100,
          "operation": {
            "operation-type": "composite",
            "requests": [
              {
                "stream": [
                  {
                    "operation-type": "raw-request",
                    "path": "/geonames/_search",
                    "body": {
                      "query": {
                        "bool": {
                          "must": {
                            "query_string": {
                              "query": "Java Performance"
                            }
                          }
                        }
                      }
                    }
                  },
                  {
                    "operation-type": "sleep",
                    "duration": 2
                  }
                ]
              },
              {
                "stream": [
                  {
                    "operation-type": "raw-request",
                    "path": "/geonames/_search",
                    "body": {
                      "query": {
                        "term": {
                          "country_code.raw": "AT"
                        }
                      }
                    }
                  }
                ]
              }
            ]
          }
        }
```

File: esrally/track/loader.py [hunk 1]
```diff
@@ -1382,13 +1382,11 @@ def parse_operation(self, op_spec, error_ctx="operations"):
             op = track.OperationType.from_hyphenated_string(op_type_name)
             if "include-in-reporting" not in params:
                 params["include-in-reporting"] = not op.admin_op
-            op_type = op.name
-            self.logger.debug("Using built-in operation type [%s] for operation [%s].", op_type, op_name)
+            self.logger.debug("Using built-in operation type [%s] for operation [%s].", op_type_name, op_name)
```
Comment (dliappis) [inline esrally/track/loader.py:1385]: Good catch.

File: esrally/track/track.py [hunk 2]
```diff
@@ -564,15 +565,20 @@ def admin_op(self):
         # pylint: disable=comparison-with-callable
         return self.value > 1000
 
+    def to_hyphenated_string(self):
+        # Pylint complains that self.name is not iterable
+        # pylint: disable=not-an-iterable
+        return "".join(["-" + c.lower() if c.isupper() else c for c in self.name]).lstrip("-")
```
Comment (dliappis) [inline esrally/track/track.py:571]: nit: took me a little to understand (by reading the method name) that we are hyphenating based on capitalization. Looking at the implementation it became clear, so not sure we need to rename, but perhaps a docstring would be useful?

**Review by dliappis — APPROVED** [review]
This looks great and will be super useful.

Left mostly quite minor docs suggestions. LGTM.

File: docs/track.rst [hunk 3]
```diff
@@ -1592,6 +1592,175 @@ This operation requires at least Elasticsearch 7.5.0 (non-OSS). This is an admin
...
+                  }
+                ]
+              }
+            ]
+          }
+        }
+      ]
+    }
+
+This executes both search requests concurrently. If both should be executed in sequence, put them in one stream. In the following example we also simulate 500ms think time in between requests::
```
[the remark this answers is not in the document]
Comment (danielmitterdorfer) [reply docs/track.rst]: Thanks. Reworded in https://github.com/elastic/rally/pull/1112/commits/fc29cc6685bd18e77824ae00eacbdad9acadb411.

File: esrally/track/track.py (hunk 2 above)
Comment (danielmitterdorfer) [reply esrally/track/track.py:574]: I added a docstring in https://github.com/elastic/rally/pull/1112/commits/fc29cc6685bd18e77824ae00eacbdad9acadb411.

Comment (dliappis) [reply esrally/track/track.py:574]: Very clear thanks!

File: docs/track.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (dliappis) [reply docs/track.rst]: This is MUCH clearer IMHO, thank you!

[the remark this answers is not in the document]
Comment (danielmitterdorfer) [reply docs/track.rst]: You're welcome. Thanks for pointing it out. :)

**Pull request merged**